### PR TITLE
Dashboard

### PR DESCRIPTION
**Pull request by andreeap00**
### Dashboard

The Dashboard available for staff is displayed. It is developed to facilitate the order management process. Authorized personnel see and update the status of the order in real time, from receipt to delivery.

- The user creates account, can update personal information and delete account.
- Admin adds, updates and deletes products data. Cloudinary is used to store the handled images.
- Archived information handling in progress. I am going to undo the changes I considered at first and implement the idea recommended during 17th's daily. - solved

### CODE REVIEW

File: app/controllers/admin/dashboard_controller.rb [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    # @orders_by_user = Order
+    #                   .where.not(status: :cart)
+    #                   .where.not(created_at: nil)
+    #                   .where(status: [Order.statuses[:pending], Order.statuses[:completed], Order.statuses[:delivered]])
+    #                   .order(status: :asc, created_at: :desc)
+    #                   .includes(:user)
+    #                   .group_by(&:user)
+
+    pending_orders = Order
+                .where.not(status: :cart)
```
Comment (andreeap00) [inline app/controllers/admin/dashboard_controller.rb]: spacing updated

File: app/controllers/admin/dashboard_controller.rb [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+                .group_by(&:user)
+
+    completed_orders = Order
+                      .where.not(status: :cart)
+                      .where.not(created_at: nil)
+                      .where(status: Order.statuses[:completed])
+                      .order(created_at: :desc)
+                      .includes(:user)
+                      .group_by(&:user)
+    delivered_orders = Order
```
Comment (andreeap00) [inline app/controllers/admin/dashboard_controller.rb]: spacing updated

File: app/views/admin/archived_orders/index.html.haml [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+      %th Status
+      %th Actions
+
+  %tbody
+    - @archived_orders.each do |archived_order|
+      %tr
+        %td= archived_order.created_at.strftime('%Y-%m-%d')
+        %td= archived_order.status
+        %td= link_to 'View Products', admin_archived_order_path
+
```
Comment (andreeap00) [inline app/views/admin/archived_orders/index.html.haml]: spacing solved

File: app/views/admin/products/edit.html.haml [hunk 4]
```diff
@@ -0,0 +1,7 @@
+- provide(:title, "Edit product")
+- provide(:button_text, 'Save changes')
+%h1 Update your product
+.row
+  .col-md-6.col-md-offset-3
+    = render 'form'
+
```
Comment (andreeap00) [inline app/views/admin/products/edit.html.haml]: spacing solved

File: app/views/admin/products/new.html.haml [hunk 5]
```diff
@@ -0,0 +1,6 @@
+- provide(:title, 'New Product')
+%h1 'New Product'
+.row
+  .col-md-6.col-md-offset-3
+    = render 'form'
+
```
Comment (andreeap00) [inline app/views/admin/products/new.html.haml]: spacing solved

File: app/views/pages/home.html.haml [hunk 6]
```diff
@@ -51,11 +53,14 @@
...
                   = form_with(url: add_to_order_order_path(id: current_user.orders.find_or_create_by(status: :cart).id, product_id: product.id), method: :post) do |f|
                     = f.submit "Add to Cart"
                     %i.fas.fa-cart-plus.icon
-              - else
-                %p Please log in to add items to your cart.
+              - elsif current_user && current_user.admin?
+                = link_to "Edit", edit_admin_product_path(product), class: "btn btn-primary"
+                = link_to 'Delete', admin_product_path(product), method: :delete, data: { confirm: 'Are you sure?' }, class: 'btn btn-danger'
+              -else
+                %p Only logged in users can add to cart.
```
Comment (andreeap00) [inline app/views/pages/home.html.haml]: spacing solved

File: app/views/users/show.html.haml [hunk 7]
```diff
@@ -16,6 +16,7 @@
           %hr
       - else
         %p No Previous Orders Available.
+      %td= button_to "Delete", user_path(@user), method: :delete, data: { confirm: "Are you sure you want to delete this user?" }
     - else
       %p Log in this account to access further information.
```
Comment (andreeap00) [inline app/views/users/show.html.haml]: spacing solved

File: app/controllers/admin/dashboard_controller.rb [hunk 8]
```diff
@@ -0,0 +1,57 @@
+class Admin::DashboardController < ApplicationController
+  before_action :authorize_admin
+
+  def index
+    pending_orders = Order
```
Comment (VictorMotogna) [inline app/controllers/admin/dashboard_controller.rb]: You have a lot of lines here that repeat themselves. You should extract them, like:
```ruby
orders = Order.where.not(status: :cart) # and all other lines that also repeat
# and then do
completed_orders = orders.where(status: :pending)
completed_orders = orders.where(status: :completed)
# etc.
```

File: app/controllers/admin/products_controller.rb [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+  end
+
+  private
+    def product_params
+      params.require(:product).permit(:title, :description, :price, :category, :vegetarian, :image)
+    end
+
+    def logged_in_user
+      unless logged_in?
+          store_location
```
Comment (VictorMotogna) [inline app/controllers/admin/products_controller.rb]: Your spacing seems off here, you are using 4 spaces instead of 2 like in the rest of the files

File: app/controllers/admin/products_controller.rb [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+          flash[:danger] = "Please log in."
+          redirect_to login_url
+      end
+    end
+
+    def admin_user
+      unless current_user && current_user.admin?
```
Comment (VictorMotogna) [inline app/controllers/admin/products_controller.rb]: I would rewrite this to:
```ruby
def admin_user
  return if current_user.nil? && !current_user.admin?

  flash[:danger] = "You don't have permission to perform this action."
  redirect_to home_path
end
```

Comment (VictorMotogna) [reply app/controllers/admin/products_controller.rb]: It is easier to read and will have the same behaviour

File: app/controllers/users_controller.rb [hunk 11]
```diff
@@ -19,14 +22,53 @@ def create
       reset_session
       log_in @user
       flash[:success] = "Welcome to the Food Universe!"
-      redirect_to @user
+      if @user.admin?
```
Comment (VictorMotogna) [inline app/controllers/users_controller.rb]: This if/else could be a 1-liner. Also, no need for the `elsif @user.user?`, because you only 2 user types. In case an extra user type comes along, you wouldn't want multiple `elsifs` anyway, you will use a switch statement

```ruby
@user.admin? ? redirect_to admin_dashboard_path : redirect_to user_path(@user)
```

File: app/controllers/users_controller.rb [hunk 12]
```diff
@@ -19,14 +22,53 @@ def create
...
+  end
+
   private
     def user_params
-      params.require(:user).permit(:name, :email, :password, :password_confirmation)
+      params.require(:user).permit(:name, :email, :password, :password_confirmation, :role)
+    end
+
+    def logged_in_user
+      unless logged_in?
```
Comment (VictorMotogna) [inline app/controllers/users_controller.rb:58]: Here as well, could do `return if logged_in?`

File: app/helpers/orders_helper.rb [hunk 13]
```diff
@@ -6,4 +6,17 @@ def calculate_order_total(order)
     end
     total
   end
-end
+
+  def order_row_class(order)
```
Comment (VictorMotogna) [inline app/helpers/orders_helper.rb:10]: This could be just `"#{order.status}-row"`

File: app/helpers/sessions_helper.rb [hunk 14]
```diff
@@ -11,16 +11,20 @@ def remember(user)
 
   def current_user
     if (user_id = session[:user_id])
-      @current_user ||= User.find(user_id)
+      @current_user ||= User.find_by(id: user_id)
```
Comment (VictorMotogna) [inline app/helpers/sessions_helper.rb:14]: Why did you change this from `User.find(user_id)`? They should have the same behaviour (unless you want to use `find_by!` which will also raise an error).

File: spec/controllers/products_controller_spec.rb [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+      allow(controller).to receive(:logged_in?).and_return(true)
+    end
+
+    it 'renders new template' do
+      get :new
+      expect(response).to render_template(:new)
+    end
+  end
+
+  describe '#create' do
```
Comment (VictorMotogna) [inline spec/controllers/products_controller_spec.rb:28]: I left a similar comment on Laurentiu's PR (copied below), he used it in a test where he created users, not products, so you will need to update the example with `product` instead of `user`

Comment (VictorMotogna) [reply spec/controllers/products_controller_spec.rb:28]: There are multiple places where you do the same action:
```ruby
post :create, params: { user: valid_attributes }
```

So what you could do here, is extract that into a `subject`. Your code will look like:
```ruby
describe 'POST #create' do
  let(:valid_create_attributes) do
    {
      name: 'John Doe',
      email: generate_unique_email('johndoe@yahoo.com'),
      password: 'password123',
      password_confirmation: 'password123'
    }
  end

  subject { post :create, params: { user: valid_attributes } }

  context 'with valid attributes' do
    it 'creates a new user' do
      expect { subject }.to change(User, :count).by(1)
    end

    # all tests where the attributes are valid
  end

  # in the context where attributes are not valid
  context 'with valid attributes' do
    # here you modify, add or remove any attribute you want, to make the final params structure to be invalid
    before { valid_create_attributes.merge!({}) }

    # add all the tests with this format of attributes, where they are invalid
  end
end
```

The `subject` is basically an action, with given params. In certain contexts, you will have a `before` action, that will modify the `valid_attributes` variable. So the subject will end up doing the same action, but with the modified `valid_attributes`. So the action (calling `post :create`) will be the same, but the params can be different

File: spec/helpers/sessions_helper_spec.rb [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+      allow(helper).to receive(:logged_in?).and_return(false)
+    end
+
+    it 'does not redirect when logged in' do
+      session[:user_id] = user.id
+      expect(response).not_to eq(:redirect_to)
+      authenticate
+    end
+  end
+end
```
Comment (VictorMotogna) [inline spec/helpers/sessions_helper_spec.rb:72]: 👀

File: spec/models/product_spec.rb [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+      expect(product.errors[:price]).to include("can't be blank")
+    end
+
+    it 'validates acceptable attributes' do
+      product = Product.new(title: 'The Most Delicious Salad Ever', price: 50.33, description: "Salads", vegetarian: :vegetarian, category: :entree)
+      expect(product).to be_valid
+    end
+  end
+
+  describe 'enums' do
```
Comment (VictorMotogna) [inline spec/models/product_spec.rb:29]: These `describe enums` could be a single describe, with multiple `it`'s, or even 2 contexts, each with 1 `it`

Comment (VictorMotogna) [reply spec/models/product_spec.rb:29]: You could also make it simpler:
```ruby
describe 'enums' do
    it { is_expected.to define_enum_for(:vegetarian).with_values(%i[non_vegetarian vegetarian]) }
    it { is_expected.to define_enum_for(:category).with_values(%i[entree second_course salad pizza dessert]) }
  end
```

File: app/helpers/sessions_helper.rb (hunk 14 above)
Comment (andreeap00) [reply app/helpers/sessions_helper.rb:14]: > Why did you change this from `User.find(user_id)`? They should have the same behaviour (unless you want to use `find_by!` which will also raise an error).

At user deletion, when `User.find(user_id) `was used, I encountered "ActiveRecord::RecordNotFound" error. By changing `find` with` find_by`, the method returns `nil` when the user with that certain ID is not in DB anymore.
